EAS local iOS build with production profile fails

### Build/Submit details page URL

_No response_

### Summary

My local iOS "production" builds started to fail on recent versions of Expo and EAS CLI.

I suspect it has something to do with my main entry file called "index.js" (package.json) - however, that wasn't a problem about 2 weeks ago on "production" profiles and isn't a problem today for latest versions of Expo/EAS if you target development profile with developmentClient: true.

### Managed or bare?

Managed

### Environment

expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.5
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 22.4.0 - /usr/local/bin/node
      npm: 10.8.1 - /usr/local/bin/npm
    Managers:
      CocoaPods: 1.15.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.5, iOS 17.5, macOS 14.5, tvOS 17.5, visionOS 1.2, watchOS 10.5
    IDEs:
      Xcode: 15.4/15F31d - /usr/bin/xcodebuild
    npmGlobalPackages:
      eas-cli: 10.1.1
    Expo Workflow: managed
    
---

✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check native tooling versions
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check npm/ yarn versions
✔ Check for issues with metro config
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that native modules do not use incompatible support packages
✔ Check for legacy global CLI installed locally
✔ Check that native modules use compatible support package versions for installed Expo SDK
✔ Check that packages match versions required by installed Expo SDK

Didn't find any issues with the project!
    

### Error output

EAS local build with "production" profile fails with the following error message:

```
[RUN_FASTLANE] › Executing expobug » Bundle React Native code and images
[RUN_FASTLANE]     Bundler cache is empty, rebuilding (this may take a minute)
[RUN_FASTLANE] ▸ ** BUILD FAILED **
[RUN_FASTLANE] ▸ The following build commands failed:
[RUN_FASTLANE] ▸ 	PhaseScriptExecution Bundle\ React\ Native\ code\ and\ images /var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/eb840d6d-644a-4121-9f78-90d18b7a2f4e/build/ios/build/Build/Intermediates.noindex/expobug.build/Release-iphonesimulator/expobug.build/Script-00DD1BFF1BD5951E006B06BC.sh (in target 'expobug' from project 'expobug')
[RUN_FASTLANE] ▸ (1 failure)
[RUN_FASTLANE] ** BUILD FAILED **
[RUN_FASTLANE] 
[RUN_FASTLANE] 
[RUN_FASTLANE] The following build commands failed:
[RUN_FASTLANE] 	PhaseScriptExecution Bundle\ React\ Native\ code\ and\ images /var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/eb840d6d-644a-4121-9f78-90d18b7a2f4e/build/ios/build/Build/Intermediates.noindex/expobug.build/Release-iphonesimulator/expobug.build/Script-00DD1BFF1BD5951E006B06BC.sh (in target 'expobug' from project 'expobug')
[RUN_FASTLANE] (1 failure)
[RUN_FASTLANE] Exit status: 65
[RUN_FASTLANE] 
[RUN_FASTLANE] +---------------------------------------+
[RUN_FASTLANE] |           Build environment           |
[RUN_FASTLANE] +-------------+-------------------------+
[RUN_FASTLANE] | xcode_path  | /Applications/Xcode.app |
[RUN_FASTLANE] | gym_version | 2.221.1                 |
[RUN_FASTLANE] | sdk         | iPhoneOS17.5.sdk        |
[RUN_FASTLANE] +-------------+-------------------------+
[RUN_FASTLANE] Looks like fastlane ran into a build/archive error with your project
[RUN_FASTLANE] It's hard to tell what's causing the error, so we wrote some guides on how
[RUN_FASTLANE] to troubleshoot build and signing issues: https://docs.fastlane.tools/codesigning/getting-started/
[RUN_FASTLANE] Before submitting an issue on GitHub, please follow the guide above and make
[RUN_FASTLANE] sure your project is set up correctly.
[RUN_FASTLANE] fastlane uses `xcodebuild` commands to generate your binary, you can see the
[RUN_FASTLANE] the full commands printed out in yellow in the above log.
[RUN_FASTLANE] Make sure to inspect the output above, as usually you'll find more error information there
[RUN_FASTLANE] [!] Error building the application - see the log above
[RUN_FASTLANE] Error: The "Run fastlane" step failed with an unknown error. Refer to "Xcode Logs" below for additional, more detailed logs.

Build failed
The "Run fastlane" step failed with an unknown error. Refer to "Xcode Logs" below for additional, more detailed logs.
npx exited with non-zero code: 1
    Error: build command failed.

```

I managed to track down the issue and narrow it to Metro Bundler:

```
Starting Metro Bundler
warning: Bundler cache is empty, rebuilding (this may take a minute)
/private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/.:0: error: Unable to resolve module /var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js from /private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/.: 
Error: Unable to resolve module /var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js from /private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/.: 

None of these files exist:
  * ../../../../../../../../../var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js(.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.mjs|.native.mjs|.mjs|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.json|.json|.ios.cjs|.native.cjs|.cjs|.ios.scss|.native.scss|.scss|.ios.sass|.native.sass|.sass|.ios.css|.native.css|.css)
  * ../../../../../../../../../var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js/index(.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.mjs|.native.mjs|.mjs|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.json|.json|.ios.cjs|.native.cjs|.cjs|.ios.scss|.native.scss|.scss|.ios.sass|.native.sass|.sass|.ios.css|.native.css|.css)
Error: Unable to resolve module /var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js from /private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/.: 

None of these files exist:
  * ../../../../../../../../../var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js(.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.mjs|.native.mjs|.mjs|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.json|.json|.ios.cjs|.native.cjs|.cjs|.ios.scss|.native.scss|.scss|.ios.sass|.native.sass|.sass|.ios.css|.native.css|.css)
  * ../../../../../../../../../var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js/index(.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.mjs|.native.mjs|.mjs|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.json|.json|.ios.cjs|.native.cjs|.cjs|.ios.scss|.native.scss|.scss|.ios.sass|.native.sass|.sass|.ios.css|.native.css|.css)
    at ModuleResolver.resolveDependency (/private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:112:15)
    at DependencyGraph.resolveDependency (/private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/node_modules/metro/src/node-haste/DependencyGraph.js:231:43)
    at /private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/node_modules/metro/src/lib/transformHelpers.js:156:21
    at Server._resolveRelativePath (/private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/node_modules/metro/src/Server.js:1023:12)
``` 

### Reproducible demo or steps to reproduce from a blank project

1. git clone https://github.com/krystianhub/ExpoEASBug
2. cd ExpoEASBug
3. npm install
4. **eas build --profile production --platform ios --local**
5. Local build fails
---
1. **eas build --profile production --platform ios**
2. EAS Cloud build is successful
---
1. **eas build --profile development --platform ios --local**
3. Local development build is successful

--- a/packages/@expo/cli/CHANGELOG.md
+++ b/packages/@expo/cli/CHANGELOG.md
@@ -22,6 +22,7 @@
 - Resolve module specifiers to posix paths for Metro. ([#29696](https://github.com/expo/expo/pull/29696) by [@byCedric](https://github.com/byCedric))
 - Ensure Metro dev server closes fully when clients are connected. ([#30067](https://github.com/expo/expo/pull/30067) by [@byCedric](https://github.com/byCedric))
 - Fix app crashing when hitting debugger breakpoint through VSCode. ([#30287](https://github.com/expo/expo/pull/30287) by [@byCedric](https://github.com/byCedric))
+- Resolve real locations of file paths when using `eas build --local`.
 
 ### 💡 Others
 

--- a/packages/@expo/cli/CHANGELOG.md
+++ b/packages/@expo/cli/CHANGELOG.md
@@ -22,7 +22,7 @@
 - Resolve module specifiers to posix paths for Metro. ([#29696](https://github.com/expo/expo/pull/29696) by [@byCedric](https://github.com/byCedric))
 - Ensure Metro dev server closes fully when clients are connected. ([#30067](https://github.com/expo/expo/pull/30067) by [@byCedric](https://github.com/byCedric))
 - Fix app crashing when hitting debugger breakpoint through VSCode. ([#30287](https://github.com/expo/expo/pull/30287) by [@byCedric](https://github.com/byCedric))
-- Resolve real locations of file paths when using `eas build --local`.
+- Resolve real locations of file paths when using `eas build --local`. ([#30340](https://github.com/expo/expo/pull/30340) by [@byCedric](https://github.com/byCedric))
 
 ### 💡 Others
 

--- a/packages/@expo/cli/src/export/embed/exportEmbedAsync.ts
+++ b/packages/@expo/cli/src/export/embed/exportEmbedAsync.ts
@@ -116,7 +116,7 @@ export async function exportEmbedBundleAndAssetsAsync(
     const bundles = await devServer.legacySinglePageExportBundleAsync(
       {
         splitChunks: false,
-        mainModuleName: options.entryFile,
+        mainModuleName: resolveRealEntryFilePath(projectRoot, options.entryFile),
         platform: options.platform,
         minify: options.minify,
         mode: options.dev ? 'development' : 'production',

